CONTRIBUTING.md

```diff
@@ -41,6 +41,39 @@ cherry-picked commit:
 $ git pull --rebase upstream master
 ```
 
+Any changes that break consensus must be behind a feature gate and must have
+a merged SIMD.
+
+All changes should have unit and integration tests that cover at least 90% of
+added code paths. These tests should run quickly and not be flaky.
+
+All changes should be stress-tested with relevant test cases, if relevant test cases
+are not present, then write them.
+
+All changes should be benchmarked and evidence posted to the PR.
+Microbenchmark results along with mainnet/testnet validator timings or profiles,
+bench-tps or other relevant integration benchmarks. Any code that adds complexity
+should be justified by a comisurate improvement in speed.
+
+All changes should be reviewed by subject matter experts.
+
+All changes should be merged to master branch first and only critical changes should
+be backported to release branches.
+
+Duplicate code should generally be avoided.
+
+Features should be activated on testnet before mainnet in the closest configuration to mainnet as possible
+Relevant metrics need to be monitored and approriate follow-up given after feature activation.
+
+Avoid “hack” or “one-off” solutions, prefer well-architected designs which are not fragile.
+
+Avoid unnecessary unwrap() or panic()
```

Comment:
We can recommend the following:
* only use `unwrap()` in cases where you can prove it will *never* panic
* in cases where panic on `unwrap()` is desirable, prefer `.expect()` 
* in code where direct user interaction can be expected, utilize anyhow::Context to propagate errors to the caller where they can be shown to the user.
* in other cases use well-defined errors that convey what happened and are easy to match on
* in performance critical sections avoid allocations on the error path

Reply:
well if we use something like `Error<T, Box<SomethingHuge>>` and the actual error case happens very rarely, then it is superior to `Error<T, SomethingHuge>` in terms of performance. But you'd have to be sure the error case is actually quite rare. So the best choice in a perf-critical section depends on measurements.

As a rule of thumb, using Error<T, SomeEnum> for error propagation in perf-critical sections is usually better, as optimizer can often mitigate the overheads of copying SomeEnum around (assuming SomeEnum does not own heap allocated objects).